### PR TITLE
Rewrite scanner in c

### PR DESCRIPTION
**Pull request by amaanq**
For portability & ease of compilation

To be honest I don't fully understand what the string was for, Menhir doesn't have a very well documented grammar (ebnf-like), so it's hard to write it more appropriately in C, so I just set a 256 char limit for strings for now..

### CODE REVIEW

**Comment by 314eter** [comment]
How will this improve portability? There are a lot of tree-sitter grammars using external C++ scanners, so every system that wants to use tree-sitter will have to be able to handle C++ anyway. Do you have an example where this grammar is currently causing problems?

I think it's best to keep the scanner as close as possible to the [OCaml scanner](https://github.com/tree-sitter/tree-sitter-ocaml/blob/master/common/scanner.h), which is using C++ too.

**Comment by amaanq** [comment]
> There are a lot of tree-sitter grammars using external C++ scanners, so every system that wants to use tree-sitter will have to be able to handle C++ anyway. 

Sure, a lot do, but it's still in the minority. It just often causes headaches for distribution systems like nvim-treesitter, or OSes like windows where compilation is a headache. My goal is to convert most, if not all, the existing C++ scanners to C, Menhir isn't the first I've PRed, and I can surely do that for OCaml as well.

I did just push a fix for an infinite loop bug if \0 (null) was detected in a string, fuzzing this is good now too.

To add on, the original c++ scanner segfaults with certain input (via fuzzing) - my current implementation fixed that.

**Comment by 314eter** [comment]
> It just often causes headaches for distribution systems like nvim-treesitter, or OSes like windows where compilation is a headache.

Do you have an example where this caused problems? Are there environments where compiling C is not a problem, but C++ is a headache? As long as there is a single tree-sitter grammar using C++, nvim-treesitter will support it, and reducing the amount of C++ will not change anything until you reduce it to zero.

> My goal is to convert most, if not all, the existing C++ scanners to C.

Are you the only one with this goal, or are there any signs that the whole tree-sitter community will follow and C++ support will be dropped?

> I did just push a fix for an infinite loop bug if \0 (null) was detected in a string, fuzzing this is good now too.

That's unrelated to the C rewrite. If you want to make it easy to review, it's better to do the fix in a separate commit or PR. If there are bugs, I think it's best to make sure they are fixed in tree-sitter-ocaml, and then sync the tree-sitter-menhir implementation (it's a bit behind right now).

> To add on, the original c++ scanner segfaults with certain input (via fuzzing) - my current implementation fixed that.

Same remark: if you fix a bug, please describe the bug and how you fixed it.

To summarize: I don't really see the benefit of rewriting it in C. Especially not if it will let the parser fail on some input because of the 256 char limit. But it would be nice to get the bugs fixed.

How did you do the fuzzing? Is that something that can be added to the repository to test future changes?

**Comment by Kerl13** [comment]
Here are my thoughts on the two (separate) issues:

- As for @314eter, my main concern is that I'd like the bugs to be fixed. And I think it would be easier to review if they were the subject of a separated pull request, with some explanations on the bugs and the fixes. Also I'd be very interested to hear how you did the fuzzing, that sounds like a very interesting addition to our tests.

- Regarding the C rewrite, honestly I'm not against it on paper, but there are some problems at the moment.
  I agree that dropping the need for a C++ toolchain would be a slight enhancement, especially since the code is not that complicated. However, @314eter has a good point that keeping our scanner close to the OCaml one seems better for maintainability. And besides, you kind of need the tree-sitter OCaml parser at the moment anyway (and hence the C++ toolchain).
  So for me that would be an okay change, but it is not worth introducing limitations to the grammar (the 256 chars bound), and it only makes sense if the people behind the ocaml parser agree the do the same move.

**Comment by 314eter** [comment]
The OCaml grammar switched to C++ to be able to use the variable-length `std::string` (tree-sitter/tree-sitter-ocaml#19). The only way to replace that in C is with a custom variable-length string implementation. The downsides of doing that outweigh the benefit of using C.

**Comment by Kerl13** [comment]
Fair enough

**Comment by amaanq** [comment]
Adjusted from https://github.com/tree-sitter/tree-sitter-ocaml/pull/78, scanner now uses variable length strings properly

The bugs that I fixed were in two spots where if a null character was detected, it wasn't advanced if it wasn't the end of the file. It was easy to demonstrate via fuzzing. hence I just added advance calls appropriately.

I can add some CI actions to fuzz this and the ocaml parser if that'd be welcome
